apps/api-documenter/src/documenters/MarkdownDocumenter.ts

```diff
@@ -763,6 +785,29 @@ export class MarkdownDocumenter {
     }
   }
 
+  private _tryCreateLinkTagForTypeExcerpt(excerpt: Excerpt): DocLinkTag | undefined {
+      const configuration: TSDocConfiguration = this._tsdocConfiguration;
+      // Use the last token, so it works for namespace types, e.g. types.MyType
+      const typeExcerptToken: ExcerptToken = excerpt.tokens[excerpt.tokenRange.endIndex - 1];
```

Comment:
@Feiyang1 you need to go loop through the individual tokens and only hyperlink the tokens with `ExcerptTokenKind.Reference`.

The DocFX/YAML documenter's [_renderType()](https://github.com/microsoft/rushstack/blob/589608e25a9a05d9f948c66c8f563f7834628760/apps/api-documenter/src/documenters/YamlDocumenter.ts#L660) is not a good example to work from, because the YAML file has a weird specialized representation for "complex" types.  Whereas for the Markdown documenter, we're just making plain text with some hyperlinks inside.

This means that `_tryCreateLinkTagForTypeExcerpt()` probably needs to return a `DocParagraph` instead of a simple `DocLinkTag`.

Reply:
Here's a sketch of how to fix that: https://github.com/microsoft/rushstack/commit/a12b04b8dec3111d7129a717d7b6487b0c0f98c3

It might need some tuning to handle different ways that whitespace can end up in the `DocCodeSpan` parts. I noticed that the original implementation was `trim()`ing them for some reason. (?) 

Hope that helps!

Reply:
Ah, that makes sense! I didn't think about composite types. Your sketch is very helpful - I copy/pasted it into this PR.

I believe the original implementation `trim()`s text for [return type](https://github.com/microsoft/rushstack/blob/master/apps/api-documenter/src/documenters/MarkdownDocumenter.ts#L754) because it allows to write the one liner that handles both empty string and only whitespace cases. It probably doesn't matter format wise, so I think we should be good without `trim()`ing.

What do you think?

Reply:
@Feiyang1 Looking at some examples, I'm thinking maybe we should switch to plain text instead of code blocks for these type excerpts.  

Consider this example generated by your current PR (using `DocCodeSpan`):

> ## f() function
> 
> An example function.
> 
> <b>Signature:</b>
> 
> ```typescript
> export declare function f(a: Apple | Banana, b: Map<Apple, number>, y: string | number, z: Symbol): Apple & Banana;
> ```
> 
> ## Parameters
> 
> |  Parameter | Type | Description |
> |  --- | --- | --- |
> |  a | [Apple](./api-documenter-test.apple.md)<code> &#124; </code>[Banana](./api-documenter-test.banana.md) | two hyperlinkable types |
> |  b | <code>Map</code><code>&lt;</code>[Apple](./api-documenter-test.apple.md)<code>, number&gt;</code> | one hyperlinkable type in a generic type |
> |  y | <code>string &#124; number</code> |  |
> |  z | <code>Symbol</code> |  |
> 
> <b>Returns:</b>
> 
> [Apple](./api-documenter-test.apple.md)` & `[Banana](./api-documenter-test.banana.md)
> 
> two hyperlinkable types as the return value

And here's how it would look with `DocPlainText`:

> ## f() function
> 
> An example function.
> 
> <b>Signature:</b>
> 
> ```typescript
> export declare function f(a: Apple | Banana, b: Map<Apple, number>, y: string | number, z: Symbol): Apple & Banana;
> ```
> 
> ## Parameters
> 
> |  Parameter | Type | Description |
> |  --- | --- | --- |
> |  a | [Apple](./api-documenter-test.apple.md) \| [Banana](./api-documenter-test.banana.md) | two hyperlinkable types |
> |  b | Map&lt;[Apple](./api-documenter-test.apple.md)<!-- -->, number&gt; | one hyperlinkable type in a generic type |
> |  y | string \| number |  |
> |  z | Symbol |  |
> 
> <b>Returns:</b>
> 
> [Apple](./api-documenter-test.apple.md) &amp; [Banana](./api-documenter-test.banana.md)
> 
> two hyperlinkable types as the return value
> 

I think the second way might actually look better.  What's your opinion?

Reply:
### Option 3: links inside `<code>`

A third possibility would be to use markdown links inside `<code>` tags, like this:
```
<b>Returns:</b>
<code>[Apple](./api-documenter-test.apple.md) & [Banana](./api-documenter-test.banana.md)</code>
```

GitHub renders it like this:

> ## f() function
> 
> An example function.
> 
> <b>Signature:</b>
> 
> ```typescript
> export declare function f(a: Apple | Banana, b: Map<Apple, number>, y: string | number, z: Symbol): Apple & Banana;
> ```
> 
> ## Parameters
> 
> |  Parameter | Type | Description |
> |  --- | --- | --- |
> |  a | <code>[Apple](./api-documenter-test.apple.md) &#124; [Banana](./api-documenter-test.banana.md)</code> | two hyperlinkable types |
> |  b | <code>Map&lt;[Apple](./api-documenter-test.apple.md)</code><code>, number&gt;</code> | one hyperlinkable type in a generic type |
> |  y | <code>string &#124; number</code> |  |
> |  z | <code>Symbol</code> |  |
> 
> <b>Returns:</b>
> 
> <code>[Apple](./api-documenter-test.apple.md) & [Banana](./api-documenter-test.banana.md)</code>
> 
> two hyperlinkable types as the return value

It's definitely not standard Markdown, but a lot of other popular markdown engines seem to support it:
- [Marked](https://marked.js.org/demo/?text=%0A%3E%20%23%23%20f()%20function%0A%3E%20%0A%3E%20An%20example%20function.%0A%3E%20%0A%3E%20%3Cb%3ESignature%3A%3C%2Fb%3E%0A%3E%20%0A%3E%20%60%60%60typescript%0A%3E%20export%20declare%20function%20f(a%3A%20Apple%20%7C%20Banana%2C%20b%3A%20Map%3CApple%2C%20number%3E%2C%20y%3A%20string%20%7C%20number%2C%20z%3A%20Symbol)%3A%20Apple%20%26%20Banana%3B%0A%3E%20%60%60%60%0A%3E%20%0A%3E%20%23%23%20Parameters%0A%3E%20%0A%3E%20%7C%20%20Parameter%20%7C%20Type%20%7C%20Description%20%7C%0A%3E%20%7C%20%20---%20%7C%20---%20%7C%20---%20%7C%0A%3E%20%7C%20%20a%20%7C%20%3Ccode%3E%5BApple%5D(.%2Fapi-documenter-test.apple.md)%20%26%23124%3B%20%5BBanana%5D(.%2Fapi-documenter-test.banana.md)%3C%2Fcode%3E%20%7C%20two%20hyperlinkable%20types%20%7C%0A%3E%20%7C%20%20b%20%7C%20%3Ccode%3EMap%26lt%3B%5BApple%5D(.%2Fapi-documenter-test.apple.md)%3C%2Fcode%3E%3Ccode%3E%2C%20number%26gt%3B%3C%2Fcode%3E%20%7C%20one%20hyperlinkable%20type%20in%20a%20generic%20type%20%7C%0A%3E%20%7C%20%20y%20%7C%20%3Ccode%3Estring%20%26%23124%3B%20number%3C%2Fcode%3E%20%7C%20%20%7C%0A%3E%20%7C%20%20z%20%7C%20%3Ccode%3ESymbol%3C%2Fcode%3E%20%7C%20%20%7C%0A%3E%20%0A%3E%20%3Cb%3EReturns%3A%3C%2Fb%3E%0A%3E%20%0A%3E%20%3Ccode%3E%5BApple%5D(.%2Fapi-documenter-test.apple.md)%20%26%20%5BBanana%5D(.%2Fapi-documenter-test.banana.md)%3C%2Fcode%3E%0A%3E%20%0A%3E%20two%20hyperlinkable%20types%20as%20the%20return%20value&options=%7B%0A%20%22baseUrl%22%3A%20null%2C%0A%20%22breaks%22%3A%20false%2C%0A%20%22gfm%22%3A%20true%2C%0A%20%22headerIds%22%3A%20true%2C%0A%20%22headerPrefix%22%3A%20%22%22%2C%0A%20%22highlight%22%3A%20null%2C%0A%20%22langPrefix%22%3A%20%22language-%22%2C%0A%20%22mangle%22%3A%20true%2C%0A%20%22pedantic%22%3A%20false%2C%0A%20%22sanitize%22%3A%20false%2C%0A%20%22sanitizer%22%3A%20null%2C%0A%20%22silent%22%3A%20false%2C%0A%20%22smartLists%22%3A%20false%2C%0A%20%22smartypants%22%3A%20false%2C%0A%20%22tokenizer%22%3A%20null%2C%0A%20%22xhtml%22%3A%20false%0A%7D&version=master) renders it correctly
- [markdown-it](https://markdown-it.github.io/#md3=%7B%22source%22%3A%22%5Cn%3E%20%23%23%20f%28%29%20function%5Cn%3E%20%5Cn%3E%20An%20example%20function.%5Cn%3E%20%5Cn%3E%20%3Cb%3ESignature%3A%3C%2Fb%3E%5Cn%3E%20%5Cn%3E%20%60%60%60typescript%5Cn%3E%20export%20declare%20function%20f%28a%3A%20Apple%20%7C%20Banana%2C%20b%3A%20Map%3CApple%2C%20number%3E%2C%20y%3A%20string%20%7C%20number%2C%20z%3A%20Symbol%29%3A%20Apple%20%26%20Banana%3B%5Cn%3E%20%60%60%60%5Cn%3E%20%5Cn%3E%20%23%23%20Parameters%5Cn%3E%20%5Cn%3E%20%7C%20%20Parameter%20%7C%20Type%20%7C%20Description%20%7C%5Cn%3E%20%7C%20%20---%20%7C%20---%20%7C%20---%20%7C%5Cn%3E%20%7C%20%20a%20%7C%20%3Ccode%3E%5BApple%5D%28.%2Fapi-documenter-test.apple.md%29%20%26%23124%3B%20%5BBanana%5D%28.%2Fapi-documenter-test.banana.md%29%3C%2Fcode%3E%20%7C%20two%20hyperlinkable%20types%20%7C%5Cn%3E%20%7C%20%20b%20%7C%20%3Ccode%3EMap%26lt%3B%5BApple%5D%28.%2Fapi-documenter-test.apple.md%29%3C%2Fcode%3E%3Ccode%3E%2C%20number%26gt%3B%3C%2Fcode%3E%20%7C%20one%20hyperlinkable%20type%20in%20a%20generic%20type%20%7C%5Cn%3E%20%7C%20%20y%20%7C%20%3Ccode%3Estring%20%26%23124%3B%20number%3C%2Fcode%3E%20%7C%20%20%7C%5Cn%3E%20%7C%20%20z%20%7C%20%3Ccode%3ESymbol%3C%2Fcode%3E%20%7C%20%20%7C%5Cn%3E%20%5Cn%3E%20%3Cb%3EReturns%3A%3C%2Fb%3E%5Cn%3E%20%5Cn%3E%20%3Ccode%3E%5BApple%5D%28.%2Fapi-documenter-test.apple.md%29%20%26%20%5BBanana%5D%28.%2Fapi-documenter-test.banana.md%29%3C%2Fcode%3E%5Cn%3E%20%5Cn%3E%20two%20hyperlinkable%20types%20as%20the%20return%20value%22%2C%22defaults%22%3A%7B%22html%22%3Atrue%2C%22xhtmlOut%22%3Afalse%2C%22breaks%22%3Afalse%2C%22langPrefix%22%3A%22language-%22%2C%22linkify%22%3Atrue%2C%22typographer%22%3Atrue%2C%22_highlight%22%3Atrue%2C%22_strict%22%3Afalse%2C%22_view%22%3A%22html%22%7D%7D) renders it correctly, but only if the "HTML tags" option is enabled
- [Kramdown](https://kramdown-sandbox.herokuapp.com/) (used by Jekyll) doesn't seem to support this at all

Right now I'm leaning towards **Option 2 `DocPlainText`**.

Reply:
@Feiyang1 here is an implementation of Option 2: https://github.com/Feiyang1/rushstack/pull/1

Reply:
I like the output of Option 3, but it's unfortunately not standard markdown. So yeah, Option 2 seems to be the most reasonable choice.

I went ahead and resolved the merge conflict in your PR. Feel free to merge.